### PR TITLE
allow custom licenses to be created and updated per version

### LINKED ISSUE #18034: implement full license selection/definition for addon submission api
follow-up for #7811 

currently you can only select from pre-defined licenses - and only by specifying a numeric license id. 

So worked needed to:
- implement the ability to create a custom license
- ~select from predefined licenses by slug or some other mnemonic~ #18361
- ~maybe make the field optional (for listed versions only?) and default to all rights reserved, or the previous license for subsequent versions~ #18367

### PR DESCRIPTION
**Pull request by eviljeff**
fixes #18034 - note you still need to know the license id unless you are specifying a custom license, which makes it a little impractical for general use (see #18361)

### CODE REVIEW

File: docs/topics/api/addons.rst [hunk 1]
```diff
@@ -482,7 +482,9 @@ This endpoint allows a submission of an upload to an existing add-on to create a
         where min/max versions from the manifest, or defaults, will be used.  See :ref:`examples <version-compatibility-examples>`.
     :<json string compatibility[app_name].max: Maximum version of the corresponding app the version is compatible with. Should only be enforced by clients if ``is_strict_compatibility_enabled`` is ``true``.
     :<json string compatibility[app_name].min: Minimum version of the corresponding app the version is compatible with.
-    :<json int license: The license id.
+    :<json int license: The id of a non-custom license. The license must match the add-on type. The license must match the add-on type. Either provide ``license`` or ``custom_license``, not both.
```
Comment (bobsilverberg) [inline docs/topics/api/addons.rst:485]: It looks like you duplicated "The license must match the add-on type."

**Review by bobsilverberg — APPROVED** [review]
This looks good, nice work @eviljeff. I may have missed it, but are there tests for cases where someone submits both a license id and a custom license? Should the API call fail in that case?

r+wc

File: docs/topics/api/addons.rst (hunk 1 above)
Comment (eviljeff) [reply docs/topics/api/addons.rst:485]: :copypasta: 😬

**Comment by eviljeff** [comment]
> This looks good, nice work @eviljeff. I may have missed it, but are there tests for cases where someone submits both a license id and a custom license? Should the API call fail in that case?

It's a bit of a edge case - because both serializer fields have the same source (`source='license'`, which is implicit for the field also named `license`) then the latter (`custom_license`) overrides the former if you specify both.  I was going to add a validation check for it but it's a little to difficult to even detect as by the time you get to `def validate` the data has already been combined.  I guess I'll call it out in the docs and add a test to confirm my assertion.

**Comment by eviljeff** [comment]
I added a check for both license and custom_license being provided.

**Pull request merged**